File: strings/levenshtein_distance.go
package main

import (
	"fmt"
)

func min(a, b, c int) int {
	if a < b {
		if a < c {
			return a
		}
		return c
	}
	if b < c {
		return b
	}
	return c
}

func levenshteinDistanceOptimized(s1, s2 string) int {
	r1 := []rune(s1)
	r2 := []rune(s2)

	if len(r1) < len(r2) {
		return levenshteinDistanceOptimized(s2, s1)
	}

	if len(r2) == 0 {
		return len(r1)
	}

	previousRow := make([]int, len(r2)+1)
	for i := range previousRow {
		previousRow[i] = i
	}

	for i, c1 := range r1 {
		currentRow := make([]int, 0)
		currentRow = append(currentRow, i+1)

		for j, c2 := range r2 {
			insertions := previousRow[j+1] + 1
			deletions := currentRow[j] + 1
			substitutions := previousRow[j]
			if c1 != c2 {
				substitutions++
			}
			currentRow = append(currentRow, min(insertions, deletions, substitutions))
		}
		previousRow = currentRow
	}

	return previousRow[len(previousRow)-1]
}

func main() {
	s1 := "kitten"
	s2 := "sitting"
	fmt.Printf("Distance between %s and %s is: %d\n", s1, s2, levenshteinDistanceOptimized(s1, s2))
}

File: strings/levenshtein_distance.cpp
/**
 * @file
 * @brief Levenshtein Distance Algorithm (Optimized)
 * @details
 * Compute the Levenshtein distance between two words (strings).
 * The function is optimized for efficiency by modifying rows in place.
 * Time Complexity: O(min(N, M)) space
 */

#include <algorithm>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

/**
 * @brief Computes Levenshtein distance
 * @param s1 First string
 * @param s2 Second string
 * @return int Distance
 */
int levenshtein_distance_optimized(std::string s1, std::string s2) {
  if (s1.length() < s2.length()) {
    return levenshtein_distance_optimized(s2, s1);
  }

  if (s2.length() == 0) {
    return s1.length();
  }

  std::vector<int> previous_row(s2.length() + 1);
  std::iota(previous_row.begin(), previous_row.end(), 0);

  for (size_t i = 0; i < s1.length(); ++i) {
    std::vector<int> current_row(s2.length() + 1);
    current_row[0] = i + 1;

    for (size_t j = 0; j < s2.length(); ++j) {
      int insertions = previous_row[j + 1] + 1;
      int deletions = current_row[j] + 1;
      int substitutions = previous_row[j] + (s1[i] != s2[j]);
      current_row[j + 1] = std::min({insertions, deletions, substitutions});
    }
    previous_row = current_row;
  }

  return previous_row.back();
}

int main() {
  std::string s1 = "kitten";
  std::string s2 = "sitting";

  std::cout << "Distance between " << s1 << " and " << s2
            << " is: " << levenshtein_distance_optimized(s1, s2) << std::endl;

  return 0;
}
